File: iomerge-server-app/src/main/java/com/github/krzychek/server/movementReader/MouseTrapReader.java
package com.github.krzychek.server.movementReader;


import com.github.krzychek.server.api.appState.AppState;
import com.github.krzychek.server.api.appState.AppStateManager;
import com.github.krzychek.server.api.movementReader.IOListener;
import com.github.krzychek.server.ui.UIHelper;
import com.github.krzychek.server.utils.BooleanCondition;
import com.github.krzychek.server.utils.IOListenerToAWTAdapter;
import org.pmw.tinylog.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;
import javax.swing.JFrame;
import java.awt.AWTException;
import java.awt.GraphicsEnvironment;
import java.awt.MouseInfo;
import java.awt.Point;
import java.awt.Rectangle;
import java.awt.Robot;
import java.util.Arrays;


/**
 * MovementReader based on transparent JFrame, catches mose inside
 */
@Component
class MouseTrapReader extends JFrame {


	private final IOListener listener;
	private final AppStateManager appStateManager;
	private final Robot robot;
	private final MoveReadingThread readingThread;

	private volatile Point center;
	private BooleanCondition isReading = new BooleanCondition();

	@Autowired
	MouseTrapReader(IOListener listener, AppStateManager appStateManager) throws AWTException {
		super("IOMerge MovementReader");

		this.listener = listener;
		this.appStateManager = appStateManager;

		readingThread = new MoveReadingThread();
		readingThread.start();

		robot = new Robot();
	}

	@PostConstruct
	private void init() {
		// UI stuff
		reposition();
		UIHelper.makeInvisible(this, true);
		setAutoRequestFocus(true);

		IOListenerToAWTAdapter listenerAdapter = new IOListenerToAWTAdapter(listener);
		// add listener
		addMouseWheelListener(listenerAdapter);
		addMouseListener(listenerAdapter);
		addKeyListener(listenerAdapter);
	}

	private void reposition() {
		Rectangle displayRect = Arrays.stream(GraphicsEnvironment.getLocalGraphicsEnvironment().getScreenDevices())
				.map(screen -> screen.getDefaultConfiguration().getBounds())
				.max((a, b) -> a.width * a.height - b.width * b.height)
				.orElseThrow(() -> new IllegalStateException("Problem with getting display dimension"));
		setLocation(displayRect.x, displayRect.y);
		setSize(displayRect.width, displayRect.height);
	}

	private void readMove() {
		if (isReading.no()) {
			Logger.warn("readMove called, but I'm not reading");
			return;
		}
		// absolute location
		Point move = MouseInfo.getPointerInfo().getLocation();
		// relative location
		move.translate(-center.x, -center.y);

		if (move.x != 0 || move.y != 0) {
			listener.move(move.x, move.y);
			centerMousePointer();
		}
	}

	private void centerMousePointer() {
		robot.mouseMove(center.x, center.y);
	}

	synchronized private void startReading() {
		if (isReading.yes()) return;

		setVisible(true);

		center = getLocation();
		center.translate(getWidth() / 2, getHeight() / 2);

		centerMousePointer();

		isReading.makeYes();
	}

	synchronized private void stopReading() {
		if (isReading.no()) return;
		isReading.makeNo();
		try {
			Thread.sleep(10);
		} catch (InterruptedException e) {
			throw new IllegalStateException(e);
		}

		appStateManager.restoreMouse();
		setVisible(false);
	}

	@EventListener
	public void onStateChange(AppState newState) {
		if (AppState.ON_REMOTE == newState) {
			startReading();
		} else {
			stopReading();
		}
	}

	@PreDestroy
	private void shutdown() {
		stopReading();
		readingThread.interrupt();
	}


	private class MoveReadingThread extends Thread {

		private volatile boolean interrupted = false;

		MoveReadingThread() {
			super("MouseTrapReader : mouse move reading thread");
		}

		@Override
		public void run() {

			while (!interrupted) {
				// await
				isReading.await();

				// read while is reading
				isReading.whileTrue(MouseTrapReader.this::readMove);
			}
		}

		public synchronized void interrupt() {
			interrupted = true;
			super.interrupt();
		}
	}

}

File: iomerge-server-app/src/main/java/com/github/krzychek/iomerge/server/movementReader/MouseTrapReader.kt
package com.github.krzychek.iomerge.server.movementReader


import com.github.krzychek.iomerge.server.api.appState.AppState
import com.github.krzychek.iomerge.server.api.appState.AppStateManager
import com.github.krzychek.iomerge.server.api.movementReader.IOListener
import com.github.krzychek.iomerge.server.utils.BooleanCondition
import com.github.krzychek.iomerge.server.utils.IOListenerToAWTAdapter
import com.github.krzychek.iomerge.server.utils.makeInvisible
import org.pmw.tinylog.Logger
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Component
import java.awt.GraphicsEnvironment
import java.awt.MouseInfo
import java.awt.Point
import java.awt.Robot
import java.util.*
import javax.annotation.PostConstruct
import javax.annotation.PreDestroy
import javax.swing.JFrame


/**
 * MovementReader based on transparent JFrame, catches mouse inside
 */
@Component
internal open class MouseTrapReader(private val listener: IOListener, private val appStateManager: AppStateManager) : JFrame("IOMerge MovementReader") {

	private val robot = Robot()
	private val readingThread = MoveReadingThread()

	@Volatile private var center = Point()
	private val isReading = BooleanCondition()

	@PostConstruct
	private fun init() {
		readingThread.start()
		// UI stuff
		reposition()
		makeInvisible()
		isAutoRequestFocus = true

		IOListenerToAWTAdapter(listener).let {
			addMouseWheelListener(it)
			addMouseListener(it)
			addKeyListener(it)
		}
	}

	private fun reposition() {
		GraphicsEnvironment.getLocalGraphicsEnvironment().screenDevices
				.map { it.defaultConfiguration.bounds }
				.maxWith(Comparator { a, b -> a.width * a.height - b.width * b.height })!!
				.let {
					setLocation(it.x, it.y)
					setSize(it.width, it.height)
					center = Point(it.x + it.width / 2, it.y + it.height / 2)
				}
	}

	private fun readMove() {
		if (isReading.isTrue.not()) {
			Logger.warn("readMove called, but I'm not reading")
			return
		}

		MouseInfo.getPointerInfo().location.apply {
			translate(-center.x, -center.y) // translte for relative location of mouse
		}.apply {
			if (x != 0 || y != 0) {
				listener.move(x, y)
				centerMousePointer()
			}
		}

	}

	private fun centerMousePointer() {
		robot.mouseMove(center.x, center.y)
	}

	@Synchronized private fun startReading() {
		if (isReading.isTrue) return

		isVisible = true

		center = location.apply {
			translate(width / 2, height / 2)
		}

		centerMousePointer()

		isReading.isTrue = true
	}

	@Synchronized private fun stopReading() {
		if (isReading.isTrue.not()) return
		isReading.isTrue = false
		try {
			Thread.sleep(10) // ensure last move was read to not center mouse after restore
		} catch (e: InterruptedException) {
			throw IllegalStateException(e)
		}

		appStateManager.restoreMouse()
		isVisible = false
	}

	@EventListener
	fun onStateChange(newState: AppState) {
		when (newState) {
			AppState.ON_REMOTE -> startReading()
			else -> stopReading()
		}
	}

	@PreDestroy
	private fun shutdown() {
		stopReading()
		readingThread.interrupt()
	}


	private inner class MoveReadingThread : Thread("MouseTrapReader : mouse move reading thread") {

		@Volatile private var interrupted = false

		override fun run() {

			while (interrupted.not()) {

				isReading.await()
				while (isReading.isTrue) readMove()
			}
		}

		override fun interrupt() {
			interrupted = true
			super.interrupt()
		}
	}

}
